app/Main.hs

```diff
@@ -0,0 +1,151 @@
+-- My lsp for some reason prefers this when formatting
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main (main) where
+
+import Control.Applicative (liftA2, (<|>))
+import Control.Monad (guard, (<=<))
+import Data.Either.Extra (maybeToEither)
+import Data.Foldable (foldl')
+import Data.List (foldl1')
+import Data.Map.Strict (Map, empty, findWithDefault, insert, (!?))
+import Data.Text qualified as Text
+import Data.Text.IO qualified as TextIO
+import System.Console.ANSI (Color (..), ColorIntensity (..), ConsoleLayer (..), SGR (..), clearScreen, setCursorPosition, setSGR)
+import Text.Read (readMaybe)
+
+type Coordinates = (Int, Int)
+
+data Player = Cross | Circle deriving (Eq, Show)
+
+showPlayerT :: Player -> Text.Text
+showPlayerT = Text.pack . show
+
+newtype Board = Board (Map Coordinates Player)
+
+showBoard :: Board -> Text.Text
+showBoard (Board board) = foldl' replaceExclamation unfilledBoard $ flip (findWithDefault " ") textBoard <$> coordinates
+  where
```

Comment:
Can you explain "syntactic bound to the outer scope"?

Reply:
Read https://wiki.haskell.org/Let_vs._Where

Reply:
Interesting but disappointing now I'll just be paranoid

Reply:
But actually in this example a where clause is okay because I use the parameter inside the clause

Reply:
But can't the compiler optimize this?
I can't really say that I understand how to write compiler optimizations but I feel like it should be possible for the compiler to notice when you use the parameters or don't use the parameters inside a where clause and move the defenition outside accordingly

And secondly is there a lint for this that you know of?

Reply:
> Interesting but disappointing now I'll just be paranoid

When you learn a new environment it's okay and even encouraged to stick with simple rules, i.e. always using `let ... in ...`, until you get enough experience to know better.


> But actually in this example a where clause is okay because ...

I honestly don't know, haven't looked much into it. I use the above approach regularly to focus on stuff I care more about.


>But can't the compiler optimize this?

Compilers can do [amazing things](https://invidious.snopyta.org/watch?v=bSkpMdDe4g4).
If you wanna watch some great GHC-oriented talks I highly recommend [these](https://invidious.snopyta.org/watch?v=uR_VzYxvbxg) [Simon Peyton Jones](https://invidious.snopyta.org/watch?v=06x8Wf2r2Mc) [talks](https://invidious.snopyta.org/watch?v=6COvD8oynmI).